File: core/SuccessorLcci/SuccessorLcci.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


def inorderSuccessor(root: TreeNode, p: TreeNode) -> TreeNode:
    queue, pre_node, cur_node = [], None, root
    while queue or cur_node:
        while cur_node:
            queue.append(cur_node)
            cur_node = cur_node.left
        cur_node = queue.pop()
        if pre_node == p:
            return cur_node
        pre_node = cur_node
        cur_node = cur_node.right

    return None

File: core/SuccessorLcci/SuccessorLcci.java
package com.dbc;

import java.util.Stack;

public class SuccessorLcci {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

    public TreeNode inorderSuccessor(TreeNode root, TreeNode p) {
        Stack<TreeNode> stack = new Stack<>();
        TreeNode preNode = null;
        while (!stack.isEmpty() || root != null) {
            while (root != null) {
                stack.add(root);
                root = root.left;
            }
            root = stack.pop();
            if (preNode == p) return root;
            preNode = root;
            root = root.right;
        }
        return null;
    }
}
